src/workerd/api/urlpattern.c++

```diff
@@ -4,289 +4,191 @@
 
 #include "urlpattern.h"
 
+#include "ada.h"
+
 #include <kj/vector.h>
 
 namespace workerd::api {
+URLPattern::URLPatternInit create(const ada::url_pattern_init& other) {
+  URLPattern::URLPatternInit result{};
+#define V(_, name)                                                                                 \
+  if (auto v = other.name) {                                                                       \
+    result.name = kj::String(kj::heapArray(v->c_str(), v->size()));                                \
+  }
+  URL_PATTERN_COMPONENTS(V)
+#undef V
 
-namespace {
-jsg::JsRef<jsg::JsRegExp> compileRegex(
-    jsg::Lock& js, const jsg::UrlPattern::Component& component, bool ignoreCase) {
-  return js.tryCatch([&] {
-    jsg::Lock::RegExpFlags flags = jsg::Lock::RegExpFlags::kUNICODE;
-    if (ignoreCase) {
-      flags = static_cast<jsg::Lock::RegExpFlags>(
-          flags | static_cast<int>(jsg::Lock::RegExpFlags::kIGNORE_CASE));
-    }
-    return jsg::JsRef<jsg::JsRegExp>(js, js.regexp(component.getRegex(), flags));
-  }, [&](auto reason) -> jsg::JsRef<jsg::JsRegExp> {
-    JSG_FAIL_REQUIRE(TypeError, "Invalid regular expression syntax.");
-  });
+  if (auto v = other.base_url) {
+    result.baseURL = kj::String(kj::heapArray(v->c_str(), v->size()));
+  }
+  return result;
 }
 
-jsg::Ref<URLPattern> create(jsg::Lock& js, jsg::UrlPattern pattern) {
-  bool ignoreCase = pattern.getIgnoreCase();
+ada::url_pattern_options URLPattern::URLPatternOptions::toAdaType() const {
+  ada::url_pattern_options options;
+  options.ignore_case = ignoreCase.orDefault(false);
+  return options;
+}
 
-  // Might look a bit confusing here. The URL_PATTERN_COMPONENTS macro
-  // is used also to define the constructor for URLPattern so to make
-  // sure things line up right we reuse that pattern here also. Because
-  // we are moving the pattern into the constructor, we need to make sure
-  // the regex patterns are compiled first so we use the macro twice.
-#define V(Name, var) auto var = compileRegex(js, pattern.get##Name(), ignoreCase);
+ada::url_pattern_init URLPattern::URLPatternInit::toAdaType() {
+  ada::url_pattern_init init{};
+#define V(_, name)                                                                                 \
+  KJ_IF_SOME(v, name) {                                                                            \
+    init.name = std::string(v.begin(), v.size());                                                  \
+  }
   URL_PATTERN_COMPONENTS(V)
 #undef V
-
-#define V(_, var) , kj::mv(var)
-  return jsg::alloc<URLPattern>(kj::mv(pattern) URL_PATTERN_COMPONENTS(V));
-#undef V
+  KJ_IF_SOME(b, baseURL) {
+    init.base_url = std::string(b.begin(), b.size());
+  }
+  return init;
 }
+URLPattern::URLPatternComponentResult create(
+    jsg::Lock& js, const ada::url_pattern_component_result& other) {
+  auto result = URLPattern::URLPatternComponentResult{
+    .input = kj::String(kj::heapArray(other.input.c_str(), other.input.size())),
```

Comment:
Likewise for double copy, there has to be better constructors for this

Reply:
yes all of these needs to change.